packages/react-utils/src/createPolymorph.ts

```diff
@@ -0,0 +1,161 @@
+> = DistributiveOmit<
+  Overwrite<Overwrite<PropsOf<T>, KeyIn<'overrides', C> extends true ? C['overrides'] : object>, P>,
+  'as' | 'ref' | (KeyIn<'omit', C> extends true ? C['omit'] : '')
+> &
+  PickOrDefault<P & IfValidAs<P>, 'as', { as?: T }> &
+  PickOrDefault<P & IfValidAs<P>, 'ref', { ref?: PropsOf<T>['ref'] }>
+
+/**
+ * Render function of `createPolymorph`.
+ *
+ * Enriches polymorph development typings, allowing default native props to be
+ * derived by the initial `as` prop (type `T`). This will not affect the
+ * consumer-facing API, only the component implementation.
+ */
+export type PolymorphicRenderFunction<
+  P extends Props,
+  T extends ElementType = AsOf<P>,
+  C extends Exact<Config<P, T>, C> = object,
+> = (props: PolymorphicProps<P, T, C>) => PolymorphicElement<P, T>
+
+/**
+ * Return type of `createPolymorph`.
+ *
+ * Enriches polymorph consumer typings, allowing changes to `as` redefine the
+ * `ElementType` and its native attributes. These changes will not affect types
+ * within component implementation; only the consumer-facing API.
+ */
+export type PolymorphicExoticComponent<
+  P extends Props,
+  T extends ElementType = AsOf<P>,
+  C extends Exact<Config<P, T>, C> = object,
+> = <TT = AsOf<P> extends ElementType ? P['as'] : T>(
+  props: PolymorphicProps<P, TT extends ElementType ? TT : T, C>,
+) => PolymorphicElement<P, TT extends ElementType ? TT : T>
+
+/**
+ * Creates a polymorphic component.
+ * This combines user-provided props `P` with derived native props of element `T`.
+ *
+ * Exposes `as` prop to allow root element, and its native attribute types, to
+ * be changed. Changing the `as` prop will also change the `ref` type. Accepts
+ * any native `Element` tag name ('div', 'input', 'button'), or a custom React
+ * component.
+ *
+ * Optionally accepts a config `C` type to fine-tune the attributes exposed
+ * to consumers and their types.
+ * - `overrides` allows modification of derived native props.
+ * - `omit` allows removal of props from the resulting polymorphic component.
+ *
+ * @param render - Render function of `PolymorphicExoticComponent`.
+ *
+ * @returns `PolymorphicExoticComponent`
+ *
+ * @example
+ * ```tsx
+ * type ButtonProps = { variant: 'success' | 'error' }
+
+ * type ButtonConfig = {
+ *   // Remove 'reset' prop from type 'button'
+ *   overrides: { type: 'button' | 'submit' },
+ *   // Disallow custom `children`
+ *   omit: 'children'
+ * }
+ *
+ * const Button = createPolymorph<
+ *   ButtonProps,
+ *   'button',
+ *   ButtonConfig
+ * >(({ as: Tag = 'button', type, variant, ...rest }) => {
+ *   return (
+ *     <Tag {...rest}>
+ *       {variant === 'success' ? 'Yaay!' : 'Oh dear...'}
+ *     </Tag>
+ *   )
+ * })
+ * ```
+ */
+export const createPolymorph = <
+  P extends Props,
+  T extends ElementType,
+  C extends Exact<Config<P, T>, C> = object,
+>(
+  render: PolymorphicRenderFunction<P, T, C>,
+) => render as PolymorphicExoticComponent<P, T, C>
```

Comment:
suggestion: Perhaps just add a console error/warning if React version is 18 or lower :)